### DISCUSSION TITLE
[react] Types for 20

### DISCUSSION THREAD

**Discussion by eps1lon** [DefinitelyTyped/DefinitelyTyped#69436] [Issues with a @types package]
React 20 is not released. This issue is meant to collect type-related breaking changes we want to ship with the next major:

1. Mark error from componentDidCatch to be `unknown`: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/69434#pullrequestreview-2020647993
2. Remove default type narrowing for `ChangeEvent.target` 
3. Remove fork for TypeScript 5.0 and below: https://github.com/DefinitelyTyped/DefinitelyTyped/discussions/69436#discussioncomment-15686540

[a comment by an automated account is not shown]

**Comment by eps1lon** [comment]
Proposal to remove `React` namespace  by @remcohaszing 

> React 19 [removes UMD builds](https://react.dev/blog/2024/04/25/react-19-upgrade-guide#umd-builds-removed). So the following line should be removed as well:
>
>```ts
>export as namespace React;
>```
>
>Unfortunately GitHub doesn’t allow placing comments on untouched lines.
>
>---
>
>Perhaps the `React` namespace should be removed too as per the DefinitelyTyped style guide. Currently ESLint rule [`@definitelytyped/export-just-namespace`](https://github.com/microsoft/DefinitelyTyped-tools/blob/main/packages/eslint-plugin/docs/rules/export-just-namespace.md) is disabled to allow this. Since this is purely stylistic, it could be done in a patch version.

-- https://github.com/DefinitelyTyped/DefinitelyTyped/pull/69022#pullrequestreview-2072140212

**Reply by remcohaszing** [reply]
These are two distinct issues. I created a small [playground](https://github.com/remcohaszing/typescript-namespace-playground) to show what means what. I hope this clarifies why

1. It’s technically incorrect to keep the line `export as namespace React` for React 19.
2. This syntax is purely stylistic and could be changed in a patch version:
   ```ts
   export = React
   namespace React {
     // …
   }
   ```

**Reply by eps1lon** [reply]
> This syntax is purely stylistic and could be changed in a patch version

PRs welcome.

**Reply by eps1lon** [reply]
> It’s technically incorrect to keep the line export as namespace React for React 19.

I don't think it was ever correct. You shouldn't magically get access to a global `React`. But a lot of the ecosystem relies on this. This needs an owner that pushes this end-to-end (DT PR, codemod, fixing all `@types/` packages before we land etc).

**Comment by eps1lon** [comment]
Proposal to remove `PropsWithChildren` by @remcohaszing

> This seems to be some counterpart to VoidFunctionComponent. Perhaps it should be removed for React 19.

-- https://github.com/DefinitelyTyped/DefinitelyTyped/pull/69022#discussion_r1610513625

**Reply by maurer2** [reply]
Hello, is this removal still planned for React 20. If so, I assume the recommended approach in React 20+ is to explicitly type the children prop as `children?: React.ReactNode`, correct? Thanks.

Cheers

**Reply by eps1lon** [reply]
I personally would lint against usage. But it's a pretty big breaking change and hard to codemod in a desired way. You should just avoid the intersection since intersections aren't free. I suspect people will just reintroduce the type internally.

**Comment by eps1lon** [comment]
Type narrowing of `event.target` for `ChangeEvent` if they're not in event handlers of relevant form elements is currently incorrect. E.g. `<div onChange={} />` should have the generic `EventTarget` in `event.target` because the change event might have bubbled up. However, `input`, `textarea` and `select` should still have the narrowed target e.g. `HTMLInputElement` in `event.target` because those elements cannot have nested elements with their own change events.

Old behavior of `ChangeEvent<T>` can be restored with `ChangeEvent<T, T>`. The first parameter is the current target whereas the second parameter is the target.

See https://github.com/DefinitelyTyped/DefinitelyTyped/pull/74383 for more context

**Comment by eps1lon** [comment]
Since TypeScript 5.0 has reached end-of-life, we can remove the fork to reduce maintenance burden. We no longer test with 5.0 anyway. React types will stop working with TypeScript 5.0.

**Comment by remcohaszing** [comment]
The types of React props of builtin HTML elements accept many props that are actually invalid. It would be nice if the types only define the props that are actually used.

Notably many elements accept useless event handlers, and void elements accept children.

[TypeScript playground](https://www.typescriptlang.org/play/?jsx=4#code/KYDwDg9gTgLgBAMwK4DsDGMCWEVwGIQQAUAlHAN4BQcNcUwMSUuR1t7APAHxvt9wcAJpgBuvfhID0kuAAEYAZwC0oMMAwqoUaHACyAQxQBPOMBHAU8ABaHBAG2BQFcQTkCYBPH1gwdkzAhwALaGJsAOgRaK4hJ8OADChgAKdvpGALzkpHBpXBQAvnnRUjxF-BwARlA8MTHkkgBUcooq4OowmtpQApW5mM76cCIQmIKm4ZEAdHAAkvAKVhBIdoIoHnD6aGjAYPBoVpjL9Lj1koU1EuQA5Jdn5wKSPZSlfBxWUM8x0k3Kqm0dOq8qnA+utBsNRmFgBFLFNZnB5otlqtPJttvBBIYAOaORYKXwAZQY0xQKEcAAkACq6AAyHwkGJQ2O0SDxRkJMGJpKglJpGXIAH1+VYYIE7AAuODXAp0miSaoSDhy8QkSh5IA)

**Reply by eps1lon** [reply]
The event handlers issue needs more investigation. Events bubble so just because a div might not be the target of a `canplay` event doesn't mean that the `canplay` event doesn't propagate through it.

**Reply by remcohaszing** [reply]
You’re right about the event propagation. I was under the impression React prevented event bubbling for some reason, but I was wrong.
